### PR TITLE
Fix healthchecks that do not persist when priority is changed

### PR DESCRIPTION
**Pull request by shulin-sq**
Commit Message: Use previously calculated healthcheck when endpoints move priority levels repeatedly. This is done by changing how mutable_cross_priority_host_map_ tracks hosts. mutable_cross_priority_host_map_ is used to track if a host already exists in the previous configuration, and its healthcheck should not be recalculated. This worked only some of the time because it would remove and then add all hosts that have changed in a priority, to a map of string (ip addr:port) to Host instance. However this did not account for when an endpoint had two Host representatives in different priorities, as is in the case when an endpoint changes priorities and there is an edge case where a host can be removed from mutable_cross_priority_host_map_ before the "should we skip activate healthchecks" logic triggers. This PR fixes that by only removing an endpoint from mutable_cross_priority_host_map_ if that removal is executed from the lowest priority. This fix does assume that memberships in priorities are always calculated starting from the lowest number to the highest.

Additional Description: 
Risk Level: Med?
Testing: see note at the bottom
Docs Changes: n/a
Release Notes: added a note to the changelog
Platform Specific Features: 
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/35243 
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### Explanation
This was a difficult bug to spot because it only happens when priority changes multiple times

For example let's consider this situation: 
- we have an endpoint A that swaps between priority 0, 1, and 2
- we assume that priority is always processed starting from 0 and going up (eg, 0, 1, 2, 3, 4... etc)
- mutable_cross_priority_host_map_ is the "list" in the situation that includes all endpoints from all priorities

When priority number goes up things are ok

0 -> 1
processing priority 0: remove A
processing priority 1: add A

1 -> 2
priority 1: remove A
priority 2: add A

but things get weird when numbers go down

2 -> 1, things are still peaceful here
priority 1: add A (but this gets ignored since A is already in the list)
priority 2: remove A (!!!)

1 -> 0, at this point the list does not include A, so any logic that checks if A exists in the cross priorty host map will fail and A will be considered as a new endpoint.

without the fix: https://gist.github.com/shulin-sq/adfb4268f5f199f054e908e3fd7afae8
with the fix:
https://gist.github.com/shulin-sq/7779a341e598d81cfaeca447b0f582d1

### Testing
* functional test described in the issue: https://github.com/envoyproxy/envoy/issues/35243
* unit test, confirmed that it fails before the fix is applied

### CODE REVIEW

**Comment by adisuissa** [comment]
Thanks for tackling this!
Can you please fix DCO and format?

**Comment by shulin-sq** [comment]
@adisuissa thanks for pointing it out!

I made some changes
- ran the format script `tools/local_fix_format.sh -all`
- added DCO
- squashed my commits
- fixed some of my comments on the new part of the unit test

File: test/extensions/clusters/eds/eds_test.cc
Comment (mpuncel) [inline test/extensions/clusters/eds/eds_test.cc:1408]: priority 2 right?

Comment (shulin-sq) [reply test/extensions/clusters/eds/eds_test.cc:1408]: yes you're right. I'll just delete this line since I don't mention it below.

File: source/common/upstream/upstream_impl.cc
Comment (shulin-sq) [inline source/common/upstream/upstream_impl.cc:991]: @adisuissa is there a preference to use find() vs [] in the project? https://github.com/abseil/abseil-cpp/blob/5a01d0f77e37493570e35aedaa95c4bcf1673c7c/absl/container/internal/raw_hash_set.h#L3361 find might be O(n) as opposed to O(1) but I saw a few other places in the codebase using find with maps as well

File: changelogs/current.yaml
Comment (adisuissa) [inline changelogs/current.yaml:5]: this should probably move down to the "bug fixes" section.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [inline source/common/upstream/upstream_impl.cc:991]: nit: create a (const) temp var holding the `addressToString(host->address())` result, and pass it here and below.

Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:991]: Can you refer me to where you saw that the time complexity of find is O(n) and of the operator is O(1)?
I think both should be O(1) average complexity, and O(n) worst-case (due to hash conflicts), but I may be mistaken.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [inline source/common/upstream/upstream_impl.cc:994]: style-nit: comments start with capital letter.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [inline source/common/upstream/upstream_impl.cc:1001]: Please convert to `if (existing_host->second->priority() >= priority) ...` and remove the `continue` part.

File: test/extensions/clusters/eds/eds_test.cc
Comment (adisuissa) [inline test/extensions/clusters/eds/eds_test.cc:1407]: I think this bug should have its specific test (and please add a comment for the regression issue at the top of the test).

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:991]: nit: const

**Review by adisuissa — COMMENTED** [review]
Thanks for tracking and fixing this issue.
Overall LGTM, modulo minor comments.

**Comment by adisuissa** [comment]
Assigning @wbpcode for senior-maintainer review.
/assign @wbpcode

File: source/common/upstream/upstream_impl.cc
Comment (shulin-sq) [reply source/common/upstream/upstream_impl.cc:991]: https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h#L3545 I don't know this code well but also couldn't find any docs that explain this definitively. I will defer to your recommendation. I would be fine with keeping it since other places use find() to access this map this is an established pattern

**Comment by shulin-sq** [comment]
@adisuissa @wbpcode I have updated the PR to address the comments from the last round.

**Comment by ggreenway** [comment]
@wbpcode gentle nudge to review

**Comment by wbpcode** [comment]
I will get some free time to review this PR this weekend. Thanks. :)

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [inline source/common/upstream/upstream_impl.cc:997]: when will the `>` happens? Or should we only do the erasing when the priority is equal?

PS: I think this is a design problem of `corss_priority_map` which I can improve in future.

**Review by wbpcode — COMMENTED** [review]
LGTM overall with only one comment. Thanks for your contribution.

/wait-any

File: source/common/upstream/upstream_impl.cc
Comment (shulin-sq) [reply source/common/upstream/upstream_impl.cc:997]: changed, thanks for spotting. I think the thing I was thinking of is what I described in the comment 

```
// An earlier priority added this host, which indicates that the host moved from
      // a higher value priority to a lower value priority. We should not remove it from the
      // cross priority map.
```

but when the operators flipped it's more obvious that we only need to handle deletions when the priorities are equal. I've also updated the comment accordingly.

Regarding the design of cross_priority_map, I did think that because it's keyed on just the endpoint address, this seems to not capture differences of metadata of the endpoint but given that I was only focused on the priority part I wasn't sure if there was any other bugs here.

Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc]: Yeah, it initially is designed for lb to search a address quickly which could tolerate minor inaccuracy. But now, I need to re-think it.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
Could you merge the main to kick the CI?

**Comment by shulin-sq** [comment]
@wbpcode done. Can you help with merging?

**Comment by shulin-sq** [comment]
@wbpcode just checking in, I did a rebase from master. did that not kick off the CI?

**Pull request merged**
